README
What is fltdj?
--------------
fltdj is a PIM program in the making. 

What does PIM mean?
-------------------
Personal Information Manager (I think), a program which manages daily notes,
appointments, alarms, to-do lists, contacts, addresses, etc. 

What can it do, in its present form? 
------------------------------------
If you are in the habit of keeping daily notes, or a diary, or a journal  of
your activities, fltdj can keep them organised and accessible. In this version
you can maintain your appointments, to-do lists, addresses and contacts. It will
alert you 10 minutes before any upcoming appointments as well.

That's nice, but will it ever do anything useful?
------------------------------------------------------
If it will, do you think I would be giving it away free?
While there is no guarantee that it will do anything useful, or anything else,
there is every probability of the version number increasing in the future as
and when I get the time to add more features (especially if anyone asks).

How do I use it?
----------------
The user interface is a nice looking calendar. You can store your thoughts
for a particular day by clicking on the appropriate date, and entering the 
data into the editor at the right. To retrieve the notes for a particular day, click
on the desired date. The same editor window displays the notes, and you can add to or
modify the text if needed.
A click on the "Appointments" tab brings up your appointment list for the 
current day. You can add and make changes to or delete any appointment.
The row of tabs gives access to options for maintaining a To-do list, a 
holidays/anniversaries list as well as names and addresses of your contacts. 
Selecting these tabs will display windows where you can enter, modify, or delete 
data.
Read the file "help.html" in the doc/ subdirectory for more details.
There's a nice clock displayed as well, so you can see what time it is. And the
current day and date is shown at the bottom of the calendar window.
fltdj is a small program, and takes up very little space on the desktop, as
well as your hard disk.

What do I need to compile it? 
-----------------------------
You need FLTK (pronounced Full-tick), the Fast Light Tool Kit, a
multi-platform library, from http://www.fltk.org.

The program has been tested on version 1.3.8 of FLTK.
This should be downloaded, compiled and installed in the default places 
BEFORE compiling fltdj.

Changes in version 0.7-rc1
--------------------------------
Some more key-bindings have been added. See the documentation.

The global search facility scans and reports separate results for the journal entries and
appointments. You can click on the results to display the journal or appointments.

Sorting of contact data, which somehow got left out in 0.7,  re-enabled.

Major changes in version 0.7-beta
---------------------------------
There have been extensive changes in the user interface. The plethora of windows
popping up all over the screen in earlier versions has been discarded in favour 
of one window with a row of tabs which provides access to the various functions. 
The documentation has been updated to reflect this.

Starting from version 0.7-beta, the program no longer uses the Fl_Editor
widget written by Robert Kesterson. The editor class provided within
FLTK is used instead. This does *not* automatically word-wrap, which means that
a carraige-return needs to be typed in when you reach the end of a line, else the
words start flowing off the right edge with a horizontal scrollbar automatically
popping up at the bottom of the editor window.

One of the reasons for substituting Robert's otherwise excellent Fl_Editor
was that syntax highlighting is possible with the internal editor. 

Not that I know what to do with this facility, at the moment...

Also starting from this version, the excellent Multitabs widget provided  
by Alexey Parshin in the FLTK bazaar has been included in the source
for fltdj. Many thanks, Alexey.

What else does fltdj use?
------------------------
fltdj uses the 'cal' utility to generate the calendar display, and the 'date'
command to show the current date. These are a part of all Linux distributions.

How do I compile and install it?
--------------------------------
First copy the file fltdj-XX-src.tar.gz to your home directory.
Type 'tar -zxvf fltdj-XX-src.tar.gz'. Press Enter. This will decompress and
extract the files to a directory called 'fltdj-XX-src', where XX stands for the
version number.

Change to that directory, 'cd fltdj-XX-src'.

Type 'make clean', if you are recompiling, just to be on the safe side.
This removes any old object files and binaries, 

To construct the dynamic executable, just put all the files in one subdirectory
and type 'make fltdj', or just 'make'.
This creates the smaller executable, which is around 0.1 MB after installation,
but which needs the required libraries at run time.

If you want the static executable, type 'make static'.
This command creates a larger executable, which is around 2 MB after installation,
but which can be copied and run on another system which does not have the required
libraries. 

The Makefile will create a "hidden" directory in your home area called ~/.fltdj.
This is where the help file (help.tdj) and your data files will be kept.
Copy any data file ending in the suffix ".tdj" from earlier versions of the
program into this directory. The program will continue to read and use them.

After becoming superuser, type 'make install'.
This will strip the debugging information from the binary, thus reducing the size
to the values written above. It will copy the binary (named fltdj) to /usr/local/bin
and allow you to run the program from any directory.
If you are not the superuser, ask the System Administrator to copy the
file 'fltdj' to /usr/local/bin.

That's it.

Contact me if there is a problem. It would be nice to hear from you even if
there isn't any. But do contact me if you find a bug, or if your compiler reports
some errors (see below).

Are there any bugs?
-------------------
Of course there are. Did you really have to ask? I have tried fixing what I 
could find. The others are left for you to discover.

There are some differences in the checking of function declarations
between different versions of the GNU C compiler.
From user feedback, I have been able to rectify some
of them.

The code for this version compiles cleanly with gcc/g++ version 11.2, which is what
I have at present, on KUbuntu 22.04.
If you have any other version of Gcc, reports of any errors you encounter while 
compiling will be highly appreciated (and fixes, too, if possible!).

Known issues in 0.7
-------------------
This means that these are bugs which will be fixed later. Small bugs, but bugs all
the same.

While searching for text globally (when the "Global Search" tab is in the
foreground), a phrase typed in the "Search for" box may lead to
spurious entries in the results window. So only a single word should be typed
when searching globally.

However, when searching inside a particular month (that's when the "Journal" tab is
visible), a phrase typed in the search box there does not seem to trouble the program.

When a particular word has been found in the text describing an appointment, the program
always displays the journal entry for that particular date when the "Show" button on the
Global Search tab is clicked. To see the result of the search, click on the Appointments
tab.

Acknowledgments, Thanks and Feedback
------------------------------------
Many, many thanks are due to:
-Robert Kesterson, for his excellent FLTK editor widget, used in earlier versions
     of fltdj
-Juraj Ziegler, who reported errors and also sent the required modifications.
-Lance Perry, who suggested putting the data files in a hidden directory in the
	user's area, and supplied the patches.
-Greg Sjaardema, who urged me to enable the program to save the notes file without
	exiting.
-Don van der Haghen, who sent error messages encountered on his system.
-Joe Krahn, who suggested that the program should check for multiple instances on startup,
	and that file access operations should be verified, in case the hidden
	directory ~/.fltdj is absent. Apparently some users may do an "rm" on it
	by mistake!
	One of his suggestions is still not implemented. It will be, Real Soon Now...
-Torquil Gault, who wanted to use parts of the fltdj code for a
	(quote) "quick & dirty GUI set date pgm for the ipaq" (unquote)
	Be my guest, the program is released under the Gnu GPL. I hope you were
	successful.
-Benjamin Long, you never told me if you succeeded in cross-compiling fltdj
	for your iPaq Handheld!
-R. K. Marwah, for valuable, live, actual user feedback, and constructive
	suggestions.
-Richard Holt, for catching the date not changing after midnight. 
-Rajesh Deshpande, for catching the errors with new versions of FLTK.
-Tom Brown, for the bug reports and error messages.
-Janet Casey for putting fltdj in the GNU software directory, and for pointing out that
	the GPL license as supplied with fltdj in not quite complete.
-Darald Bantel, just for mailing me that he could successfully install fltdj.
-Leslie H. Watter, Jeremy M. Dolan, Luke Th. Bullock, Szabo Peter and SG for
	alerting me about the broken links on the web page containing
	the download files.
-Alexey Parshin for the excellent Multitabs widget.

What is the copying policy?
---------------------------
While it is copyright, (C) 2001 onwards, Kartik Patel, the program is also free
software; you can redistribute it and/or modify it under the terms of the GNU
General Public License as published by  the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.  This program
is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.  You
will receive a copy of the GNU General Public License along with this program;
for more information, write to the Free Software  oundation, Inc., 675 Mass
Ave, Cambridge, MA 02139, USA. 

What else?
----------
That covers it, more or less. Email me if there is anything left out. 
Enjoy!

Kartik Patel
letapk@gmail.com
https://letapk.wordpress.com


ChangeLog
19 June 2022 - Updated documentation.
17 Apr 2004 - Modified global search code to display journal or appointment on clicking the
			results list. Deleted "Show" button.
			Modified html doc file.
			Rebuilt 0.7rc1.
16 Apr 2004 - Sorting of contacts was broken in 0.7. Added sort_contacts while saving contact
			data.
			Display of note after global search mangles journal data. Fixed.
			Changed format of list entry after global search.
4 Apr 2004 - Added code in global search to get filename list in .fltdj directory, read and search
			the data and display results under global search tab in appt and noting column.
			Fixed a potential segfault while reading unallocated pointer (noting[0]) in journal
			file.
28 Mar 2004 - Splitting global search to a separate file.
21 Mar 2004 - Increased buffer for global search to fix segfault when lines are too
			long! Added line in searchlistshow_cb to jump to first instance of text
			found during global search.
			Buggy issue: phrase in global search leads to incorrect results. Traced to
			binary data associated with appointments, misleading grep to return wrong
			results ("Binary file matches").
			Changed name of lockfile to exclude it from global searches. The "-" was
			the culprit. Remember not include it in any  data filename.
			Fixed a strange bug in which the fortune command line was being displayed
			in the appointment number box. How things interact in weird ways. This was
			due to the commontext buffer being filled with the fortune command when
			the apptab was in the forground, and since the appt number also used the
			same buffer, this got displayed when an empty note was clicked on the
			calendar, and when the fortune option was turned on...
			Fixed.
			Remade targz package for rc1.
15 Jan 2004 - Made targz package for rc1. Not released yet.
               Modified makefile to link static executable. Needs two more
               libraries. Unzipped stripped binary is 1.8MB, as opposed to
               0.1MB for dynamic executable. Modified README. This is not in
               rc1.
25 Dec 2003 - Merry Christmas and God bless us everyone.
               Added keyboard shortcuts. Modified the documentation.
23 Dec 2003 - Added code to searchlistshow_cb to read file name and display
               journal tab. Now we need to jump to date containing the search text.
               Modified documentation.
22 Dec 2003 - Added code to searchlistshow_cb to read searchlist entry and create
               file name to read.
21 Dec 2003 - Deleted remnants of cancel button on Preferences tab. Modified the
               documentation to remove references to old behaviour.
               Added shortcut to exit button.
               Changed searchlist browser to Hold_Browser. Added show button to allow
               display of selected month. Added callback searchlistshow_cb to
               show button.
15 Oct 2003 - Modified datebutton fn in fltdjcal, to show note even when notes tab is in the
               background. The notes were being overwritten otherwise.
31 Aug 2003 - Modified display of note after search. Currentdaybuttonindex was not
			being assigned.
23 Aug 2003 - Program now opens with the About tab. On setting preferences
			to show the fortune, program will now save it (unless it is
			erased) even if nothing is typed into the notes.
			Made minor changes to documentation
17 Aug 2003 - Moved notes functions to separate file. Altered makefile to match.
			Changing the month when the notes tab was not displayed erased
			the first note of the new month. Fixed.
27 July 2003 - GlobalSearch function erases notes in current month.
			Removed old Fl_Editor dependencies from project files.
26 July 2003 - Debugged search function with respect to notes. Removed
			redundant data saves.
			When no findtext is entered, no search is done.
5 July 2003 - Changed editor for contacts to FLTK editor. Modified code to
			read and save contact data in fltdjcontacts.cpp.
			As before, slo-o-ow-ly things move.
8 June 2003 - Fixed. The editor was trying to display a note which pointed to a
			NULL pointer!
			Modified code in save_note_to_buffer.
2 June 2003 - Slo-o-ow-ly things move. Changed editor for notes from fl_editor
		to inbuilt FLTK editor. Compiles. Segfaults! To fix.
27 Apr 2003 - Changed editor for todolist from fl_editor to inbuilt FLTK editor.
		Works, except for folding of line at edge of window. Users will
		just have to press return at end of line. This also does not add one
		extra line at end of file.
24 Mar 2003 - Added statement to redraw date button after adding an appointment.
		This was omitted earlier and buttons were not being shown
		depressed until redrawn later!
		Added a dummy Fl_Help Dialog, which is loaded with the help
		file, but not shown. This allows links and images to be visible
		in the helptab.
23 Mar 2003 - Added display of holiday and current date to appt tab.
		Adjusted user interface widgets. Tweaked code relating to
		display of day/date.
22 Mar 2003 - Fixed tab switching bug by calling getdayname from inside shownote.
		Tweaked initialisation code to display holiday buttons at start of
		program.
		Added holiday textbox to journal tab.
21 Mar 2003 - Tweaked code for global search and contact search in fltdjsearch
		instead of fltdjcontacts.
		To fix bug which sometimes erases notes when switching from apptab to
		journaltab.
		Error/Feature: Todo adds one blank line to itself every time
					when exiting!
17 Mar 2003 - Search code checks for foreground tab while searching
		for strings.
16 Mar 2003 - Edited help file. Image is not being displayed in help tab.
14 Mar 2003 - Activated controls which go to next or previous day with
		appointments on the appointments tab. Tweaked the UI a little.
		Added code to show correct preferences.
13 Mar 2003 - Added code to highlight current date and display associated note
		at program startup. Added shownote and getdayname functions.

fltdjcal.cpp
/***************************************************************************
                          fltdjcal.cpp  -  description
                             -------------------
    begin                : Tue Mar 8 2003
    copyright            : (C) 2003 by Kartik Patel
    email                : letapk@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

//Last modified - 27 July, 2003

#include <unistd.h>//for unlink
#include <stdlib.h>//for getenv
#include <string.h>//for getenv
#include "fltdj.h"

void getdayname (int i);

struct HOLIANN {
	char date[3];
	char month[3];
	char description[26];
};

//data variables for the calendar================================================
//text used on labels of calendar date buttons
//38 buttons of 3 chars each; useage starts from 1, not 0
char daytext[38][3];
//text representation of the year
char yearname[6];
//text representation of name of month, which is what its name is
char monthname[11];
//text representation of name of weekday, which is what its name is
char dayname[11];
//text representation of day, date and year, used in label on note window
char notedaydate[35];

//text to hold today's day, date, and month from the output of date command
char today[15];
char todaydate[3];
char todayday[4];
char todaymonth[4];

//numerical representation of today's date
int today_date;

//this is the number of unused daybuttons in the first week
int offset;
//needed for month.tdj
int status;
//numerical representation of month
int month;
//numerical representation of year
int year;
//numerical date of the month associated with the selected daybutton
int date;
//this is the array of buttons used in the calendar; the date of the month
//associated with any button of index i is given by daybutton[i]
//use of this array starts from 1, not 0
int daybutton[38];
//flag to indicate leap year
int leap;
//this is the index of currently selected day button, 0 to 38
int currentdaybuttonindex;
//=================================================

//data variables for file I/O====================================================
extern char notefilename[128];
extern char fortunefilename[128];
//=================================================

extern Fl_Text_Buffer *notebuf;//in fltdjnotes.cpp

//these are declared and defined in fltdjappt.cpp
extern int appt[];
extern int maxapptindex;
extern int apptindex;

//these are declared and defined in fltdjholidays.cpp
extern int maxholindex;
extern struct HOLIANN holiann[101];

//these are declared and defined in fltdjprefs.cpp
extern int showfortune;

//these are declared and defined in fltdjmain.cpp
extern char commontext[];
extern int attach[];
extern char *noting[];
extern int modifiedfile;
extern FILE *holiannfile;
extern FILE *fortunefile;
extern void shownote (void);	

//callback for year and month up-down buttons
//notes data file is changed here
void cal (Fl_Widget *obj, void *data)
{
int i;
Fl_Button *but;

	save_data ();
	//clear the notes buffer
 	//notebuf->text ("\0");
	
	but = (Fl_Button *) obj;
	
	if (but == monthupbutton) {
		month++;
	}
	if (but == monthdownbutton) {
		month--;
	}

	//check if month crosses year boundary (Jan or Dec), if so adjust year
	if (month == 13) {
		month = 1;
		year++;
	}
	if (month == 0) {
		month = 12;
		year--;
	}
		
	if (but == yearupbutton) {
		year++;
	}
	if (but == yeardownbutton) {
		year--;
	}
	
	//check if year is out of range
	if (year == 10000) {
		year--;
	}
	if (year == 0) {
		year++;
	}
	
	get_cal ();
	
	for (i = 0; i < 10; i++) {
		appttime[i]->label("");
		apptname[i]->label("");
		
		appttime[i]->redraw();
		apptname[i]->redraw();
	}

 	//make the filename to be read	
	sprintf (notefilename, "%s/.fltdj/%i-%i.tdj", getenv ("HOME"), year, month);	
 	//read notes and appointments for this month
	open_journal_file ();	
 	//set the correct dates on the calendar buttons
	set_labels ();
 	//show the holidays for this month
     show_hol_buttons ();	
 	//make the first of the month the date to be displayed
 	date = 1;
 	//label it red, others black
  	//the date associated with calendar button day[i] is given by daybutton[i]
	for (i = 1; i <= 37; i++) {//scan the calendar button array
		day[i]->labelcolor(FL_BLACK);//set button color
       	if (date == daybutton[i]) {
       		day[i]->labelcolor(FL_RED);//set this button color
			day[i]->redraw ();
			getdayname(i);
  		}
	}

	shownote();
	if (basetab->value() == apptab) {
		maxapptindex = appt[date];
		apptindex = 1;
		fill_appointment_boxes (maxapptindex);		
		edit_appointment (apptindex);
	}
}

//end cal

//callback for calendar date buttons, to show note or appointments
void datebutton (Fl_Widget *obj, void *data)
{
int i, j, rem;
Fl_Button *but;

	save_note_in_buffer ();

 	//show the journaltab if some other tab (except apptab) is visible
	if (basetab->value() != journaltab && basetab->value() != apptab)
		basetab->value(journaltab);

	//color all the dates black
 	for (i = 1; i < 38; i++) {
		day[i]->labelcolor(FL_BLACK);
		day[i]->redraw();
	}
	//color the present one red
	but = (Fl_Button *) obj;
	but->labelcolor(FL_RED);
	//get the date and day name
	for (i = 1; i < 38; i++) {
		if (but == (day[i])) {
			getdayname (i);
			break;
		}
	}
	
	if (basetab->value() == apptab) {
		maxapptindex = appt[date];
		apptindex = 1;
		fill_appointment_boxes (maxapptindex);		
		edit_appointment (apptindex);
	}
     //this is to load the correct note in the buffer when the appt tab is visible
     shownote ();
}

//end datebutton

//assign the correct labels to all the calendar buttons
int set_labels (void)
{
int i;
	
	currentdate->label (today);

	//assign the dates for the first 7 buttons
	for (i = 1; i < 8; i++) {
		daybutton[8 - i] = daybutton[8] - i;
	}
	//assign the dates from 9th button onwards
	for (i = 1; i < 30; i++) {
		daybutton[8 + i] = daybutton[8] + i;
	}

	for (i = 1; i < 38; i++) {
		//remove the date boxes for illegal dates
		if ((daybutton[i] <= 0) || (daybutton[i] > 31)) {
			(day[i])->hide();
			continue;
		}
		if (month == 2 && leap == 0 && daybutton[i] > 28) {
			(day[i])->hide();
			continue;	//February, non-leap-year
		}
		if (month == 2 && leap == 1 && daybutton[i] > 29) {
			(day[i])->hide();
			continue;	//February, leap-year
		}
		if ((month == 4 || month == 6 || month == 9 || month == 11) \
			&& daybutton[i] > 30) {
			(day[i])->hide();
			continue;	//Apr, Jun, Sep, Nov - 30 days
		}
	
		//set label and boxtype for legal dates
		sprintf (daytext[i], "%i", daybutton[i]);
		(day[i])->label(daytext[i]);
		
		//square, round, up or down button type
		set_button_type (i);
		
		(day[i])->redraw();
		(day[i])->show();
	}

	//set year and month label
	monthtext->label(monthname);
	sprintf (yearname, "%i", year);
	monthtext->redraw();
	yeartext->label(yearname);
	yeartext->redraw();
	offset = 1 - daybutton[1];
	return 0;
}

//end set_labels

void set_button_type (int i)
{
	//normally the day buttons are up and square
	(day[i])->box(FL_UP_BOX);
	//unless the day has an attached note - round up button
	if (attach[daybutton[i]])
		(day[i])->box(FL_ROUND_UP_BOX);
	//unless the day has an appointment - down square button
	if (appt[daybutton[i]])
		(day[i])->box(FL_DOWN_BOX);
	//unless the day has both - round down button
	if (attach[daybutton[i]] && appt[daybutton[i]])
		(day[i])->box(FL_ROUND_DOWN_BOX);
	(day[i])->redraw();
}

//end set_button_type

void show_hol_buttons (void)
{
//Remember:
//the date associated with calendar button day[i] is given by daybutton[i]
int i, j;

	for (i = 1; i <= 37; i++) {//scan the calendar button array
		day[i]->color(FL_GRAY);//reset button color
		day[i]->tooltip("");
		day[i]->redraw ();
	}

	for (j = 1; j <= maxholindex; j++) {
		if (month == atoi (holiann[j].month)) {//for this month which is displayed on calendar
			for (i = 1; i <= 37; i++) {//scan the calendar button array
				if (daybutton[i] == atoi (holiann[j].date)) {//found a holiday!
					day[i]->color(FL_YELLOW);//color it yellow
					day[i]->tooltip(holiann[j].description);
					day[i]->redraw ();//redraw it
				}
			}
		}
	}
}

//end show_hol_buttons

//reads the calendar (from the 'cal' utility), gets the month and year,
//asigns the dates to the daybutton array, and labels to the calendar day buttons
//Don't confuse the (int) daybutton array with the fltk widget array of buttons
//called 'day'
int get_cal (void)
{
FILE *cal;
int i, j;

	//first clear all date-labels and reset daybuttons to 0
	for (i = 1; i < 38; i++) {
		daybutton[i] = 0;
		(day[i])->type ();
		(day[i])->label ("");
		//(day[i])->labelcolor (FL_BLACK);
 	}
	strncpy (monthname, "\0", 11);

	//invoke the external 'cal' utility and create the monthly calendar
	if (month == 0) {//program has just started, so current month
		status = system ("gcal > month.tdj");
		date = atoi (todaydate);//date at program start
		today_date = date;
	}
	else {//the month which has been chosen
		strcpy (commontext, "\0");
		sprintf (commontext, "gcal %i %i > month.tdj", month, year);
		status = system (commontext);
	}

	//open the calendar output file
	cal = fopen ("month.tdj", "r");
	//read the month and the year
	fscanf (cal, "%s %i\n", monthname, &year);
	//read and discard the dayname line
	fgets (commontext, 25, cal);
	//read and discard the dates of the first week
	fgets (commontext, 25, cal);
	//find the date of first Sunday of second row in calendar
	fscanf (cal, "%i\n", &(daybutton[8]));
	fclose (cal);

	leap = 0;
	//set leap year - if century
	if (((year % 100) == 0) && ((year % 400) == 0))
		leap = 1;
	//if plain leap year
	else if ((year % 4) == 0)
		leap = 1;

	//set the month
	if (strcmp (monthname, "January") == 0)
		month = 1;
	if (strcmp (monthname, "February") == 0)
		month = 2;
	if (strcmp (monthname, "March") == 0)
		month = 3;
	if (strcmp (monthname, "April") == 0)
		month = 4;
	if (strcmp (monthname, "May") == 0)
		month = 5;
	if (strcmp (monthname, "June") == 0)
		month = 6;
	if (strcmp (monthname, "July") == 0)
		month = 7;	
	if (strcmp (monthname, "August") == 0)
		month = 8;
	if (strcmp (monthname, "September") == 0)
		month = 9;
	if (strcmp (monthname, "October") == 0)
		month = 10;
	if (strcmp (monthname, "November") == 0)
		month = 11;
	if (strcmp (monthname, "December") == 0)
		month = 12;

	//delete the calendar output file
	i = unlink ("month.tdj");	

	return 0;
}

//end get_cal

void get_date (void)
{
FILE *cal;
int i, j;

	//invoke the date utility, create the file containing today's date
	status = system ("date > today.tdj");
	//open it
	cal = fopen ("today.tdj", "r");
	//read first 11 chars of the single line in the file; close it
	fgets (today, 11, cal);
	strcat (today, "\0");
	fclose (cal);
	
	//first 3 chars contain the day
	for (i = 0; i < 3; i++) {
		todayday[i] = today[i];
	}
	strcat (todayday, "\0");
	//leave a blank, then next 3 chars for the month
	for (i = 4, j = 0; i < 7; i++, j++) {
		todaymonth[j] = today[i];
	}
	strcat (todaymonth, "\0");
	//leave a blank, then next 2 chars the date
	for (i = 8, j = 0; i < 10; i++, j++) {
		todaydate[j] = today[i];
	}
	strcat (todaydate, "\0");
	//ignore the rest (time, timezone, and year); delete the file
	i = unlink ("today.tdj");
}

//end get_date

void check_date (void *j)
{
	get_date ();
	currentdate->label (today);
	Fl::repeat_timeout(600.0, check_date);
}

//end check_date

int make_a_fortune (void)
{
	strcpy (commontext, "\0");
	//make the string for the system command
	sprintf (commontext, "fortune > %s", fortunefilename);

	//issue system command and create the fortune file
	status = system (commontext);
	if (status < 0)
		return 0;//how unfortunate!
	else
		return 1;
}

//end make_a_fortune

void getdayname (int i)
{
int rem;

	date = daybutton[i];
	currentdaybuttonindex = i;

	rem = i % 7;
	//compare the remainder with the numerical day-of-week to get the dayname
	if (rem == 1)
		strcpy (dayname, "Sunday");
	if (rem == 2)
		strcpy (dayname, "Monday");
	if (rem == 3)
		strcpy (dayname, "Tuesday");
	if (rem == 4)
		strcpy (dayname, "Wednesday");
	if (rem == 5)
		strcpy (dayname, "Thursday");
	if (rem == 6)
		strcpy (dayname, "Friday");
	if (rem == 0)
		strcpy (dayname, "Saturday");

 	//display the weekday, date and month on the Journal and Appt tabs
	strcpy (notedaydate, "");
	sprintf (notedaydate, "%s %i %s %i", dayname, date, monthname, year);
  	notedaydatedisplayed1->label (notedaydate);
  	notedaydatedisplayed2->label (notedaydate);

	return;
}

//end getdayname

//end file
